include/zephyr/zbus/zbus.h

```diff
@@ -0,0 +1,507 @@
+#else
+#define ZBUS_USER_DATA_INIT /* No user data */
+#endif
+
+#if defined(CONFIG_ZBUS_RUNTIME_OBSERVERS_REGISTRATION)
+#define ZBUS_RUNTIME_OBSERVERS_LIST_INIT                                                           \
+	.runtime_observers = SYS_SLIST_STATIC_INIT(&_name.runtime_observers),
+#else
+#define ZBUS_RUNTIME_OBSERVERS_LIST_INIT /* No user data */
+#endif
+
+#if defined(CONFIG_ZBUS_STRUCTS_ITERABLE_ACCESS)
+#define _ZBUS_STRUCT_DECLARE(_type, _name) STRUCT_SECTION_ITERABLE(_type, _name)
+#else
+#define _ZBUS_STRUCT_DECLARE(_type, _name) struct _type _name
+#endif /* CONFIG_ZBUS_STRUCTS_ITERABLE_ACCESS */
+
+#define _ZBUS_OBS_EXTERN(_name) extern struct zbus_observer _name
+
+#define _ZBUS_CHAN_EXTERN(_name) extern struct zbus_channel _name
+
+#define ZBUS_REF(_value) &(_value)
+
+k_timeout_t _zbus_timeout_remainder(uint64_t end_ticks);
+/** @endcond */
+
+/**
+ * @def ZBUS_OBS_DECLARE
+ * This macro list the observers to be used in a file. Internally, it declares the observers with
+ * the extern statement. Note it is only necessary when the observers are declared outside the file.
+ */
+#define ZBUS_OBS_DECLARE(...) FOR_EACH(_ZBUS_OBS_EXTERN, (;), __VA_ARGS__)
+
+/**
+ * @def ZBUS_CHAN_DECLARE
+ * This macro list the channels to be used in a file. Internally, it declares the channels with the
+ * extern statement. Note it is only necessary when the channels are declared outside the file.
+ */
+#define ZBUS_CHAN_DECLARE(...) FOR_EACH(_ZBUS_CHAN_EXTERN, (;), __VA_ARGS__)
+
+/**
+ * @def ZBUS_OBSERVERS_EMPTY
+ * This macro indicates the channel has no observers.
+ */
+#define ZBUS_OBSERVERS_EMPTY
+
+/**
+ * @def ZBUS_OBSERVERS
+ * This macro indicates the channel has listed observers. Note the sequence of observer notification
+ * will follow the same as listed.
+ */
+#define ZBUS_OBSERVERS(...) __VA_ARGS__
+
+/**
+ * @brief Zbus channel definition.
+ *
+ * This macro defines a channel.
+ *
+ * @param _name The channel's name.
+ * @param _read_only Flag indicates the channel is read-only.
+ * @param _type The Message type. It must be a struct or union.
+ * @param _validator The validator function. It must be the type ``bool (*validator)(void *msg,
+ * size_t msg_size)``.
+ * @see struct zbus_channel
+ * @param _observers The observers list. The sequence indicates the priority of the observer. The
+ * first the highest priority.
+ * @param _init_val The message initialization.
+ */
+#define ZBUS_CHAN_DEFINE(_name, _read_only, _type, _validator, _observers, _init_val)              \
+	static _type _CONCAT(_zbus_message_, _name) = _init_val;                                   \
+	K_MUTEX_DEFINE(_CONCAT(_zbus_mutex_, _name));                                              \
+	FOR_EACH_NONEMPTY_TERM(_ZBUS_OBS_EXTERN, (;), _observers)                                 \
+	_ZBUS_STRUCT_DECLARE(zbus_channel, _name) = {                                              \
```

Comment:
Most fields of this structure is not supposed to be modified at runtime (I assume the only exception is `runtime_observers`; or am I wrong and the intention is to allow also modification of `user_data` or `read_only`?), why not try to make it `const` and locate in ROM?